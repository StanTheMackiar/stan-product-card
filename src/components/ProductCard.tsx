import { useProduct } from "../hooks/useProduct";
import React, { createContext, CSSProperties } from "react";

import {  ProductContextProps, Product, onChangeArgs, initialValues, ProductCardHandlers} from "../interfaces/interfaces";

import styles from "../styles/styles.module.css";

export const ProductContext = createContext({} as ProductContextProps);

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: ( args: ProductCardHandlers ) => JSX.Element,
  className?: string;
  style?: CSSProperties;
  onChange?: ( args: onChangeArgs ) => void;
  value?: number;
  initialValues?: initialValues;
}

export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: Props) => {

  const { counter, increaseBy, maxCount, isMaxCountReached, reset } 
  = useProduct({ onChange, product, value, initialValues });

  return (
    <ProductContext.Provider
      value={{
        counter,
        increaseBy,
        maxCount,
        product,
      }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {
          children({
            count: counter,
            isMaxCountReached,
            maxCount,
            product,

            increaseBy,
            reset,
          })
        }
      </div>
    </ProductContext.Provider>
  );
};

