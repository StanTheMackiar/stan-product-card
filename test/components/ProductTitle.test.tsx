import React from "react";
import ProductCard, { ProductTitle } from "../../src/components";
import renderer from 'react-test-renderer';
import { product1, product2 } from "./data/products";
;



describe('ProductTitle', () => {

    test('Debe mostrar el componente correctamente con el titulo personalizado', () => {

        const wrapper = renderer.create(
            <ProductTitle title="Custom product"/>
        )

        expect( wrapper.toJSON()).toMatchSnapshot()
    })


    test('debe de mostrar el componente con el nombre del producto 1', () => {

        const wrapper = renderer.create(
            <ProductCard product={ product1 }>
                {
                    () => (
                        <ProductTitle/>
                    )
                }
            </ProductCard>

            
        )

        expect( wrapper.toJSON()).toMatchSnapshot()

    })


    test('debe de mostrar el componente con el nombre del producto 2', () => {

        const wrapper = renderer.create(
            <ProductCard product={ product2 }>
                {
                    () => (
                        <ProductTitle/>
                    )
                }
            </ProductCard>

            
        )

        expect( wrapper.toJSON()).toMatchSnapshot()

    })
})