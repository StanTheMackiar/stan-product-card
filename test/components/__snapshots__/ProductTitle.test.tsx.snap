// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle Debe mostrar el componente correctamente con el titulo personalizado 1`] = `
<span
  className="productDescription undefined"
>
  Custom product
</span>
`;

exports[`ProductTitle debe de mostrar el componente con el nombre del producto 1 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Cofee Mug - Sin imagen
  </span>
</div>
`;

exports[`ProductTitle debe de mostrar el componente con el nombre del producto 2 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Cofee Mug - Con Imagen
  </span>
</div>
`;
